App/Nevermore.lua

```diff
@@ -15,49 +15,46 @@ assert(script.Name == "Nevermore", "Invalid script name. For Nevermore to work c
 assert(script.Parent == ReplicatedStorage, "Invalid parent. For Nevermore to work correctly, it should be " ..
 	"a ModuleScript named \"Nevermore\" parented to ReplicatedStorage")
 
---- Handles yielded operations by caching the retrieval process
-local function handleRetrieving(retrieving, func, argument)
-	assert(type(retrieving) == "table", "Error: retrieving must be a table")
-	assert(type(func) == "function", "Error: func must be a function")
-
-	local bindableEvent = Instance.new("BindableEvent")
-	local result
-	retrieving[argument] = function()
-		if result ~= nil and result then
-			return result
-		end
+local AsyncMemoizer = {}
+AsyncMemoizer.ClassName = "AsyncMemoizer"
+AsyncMemoizer.__index = AsyncMemoizer
 
-		bindableEvent.Event:Wait()
-		return result
-	end
+--- Caches single argument of an asyncronious function
+function AsyncMemoizer.new(func)
+	assert(type(func) == "function", "func must be a function")
+	local self = setmetatable({}, AsyncMemoizer)
 
-	result = func(argument)
-	assert(result ~= nil, "result cannot be nil")
-	retrieving[argument] = nil
-	bindableEvent:Fire()
-	bindableEvent:Destroy()
+	self._func = func
+	self._cache = {}
+	self._pending = {}
 
-	return result
+	return function(arg)
+		return self:_pend(arg)
+	end
 end
 
---- Caches single argument, single output only
-local function asyncCache(func)
-	assert(type(func) == "function", "Error: func must be a userdata")
-
-	local cache = {}
-	local retrieving = {}
+function AsyncMemoizer:_pend(arg)
```

Comment:
Maybe an assert would be good here to ensure `arg` is not nil.

```lua
assert(arg ~= nil, "AsyncMemoizer does not support nil arguments");
```

Though probably from a usability standpoint, you'd want this to not call itself an AsyncMemoizer, but "LoadLibrary", "GetRemoteEvent", or "GetRemoteFunction". Dunno, food for thought. It seems nasty to throw that in as another parameter for the constructor, but perhaps the constructor could return both the object, and the function? Or the function could be a (loosely-enforced) read-only parameter of the object?

```lua
local loadLibraryObj = AsyncMemoizer.new(debugLoading(getLibraryLoader(libCache)))
loadLibraryObj.DebugName = "LoadLibrary"
lib.LoadLibrary = loadLibraryObj.Call
```

Reply:
This is a great idea. Personally I think the `GetRemoteEvent` and `GetRemoteFunction` need to go into a helper library (and because of how Nevermore works, this will not break backwards compatability). 

Construct shouldn't probably return two things? I'm not sure. Maybe an :apply() method. Would have to look back at the context of the code.